app/api/logging/audit.py

```diff
@@ -20,54 +20,117 @@
-        # Gunicorn generates a high volume of these events in normal operation (see workertmp.py)
-        return
-    if event_name == "open" and isinstance(args[0], str) and "/pytz/" in args[0]:
-        # The pytz module generates a high volume of these events in normal operation.
+def init() -> None:
+    """Initialize the audit logging module to start
+    logging security audit events."""
+    sys.addaudithook(handle_audit_event)
+
+
+def handle_audit_event(event_name: str, args: tuple[Any, ...]) -> None:
+    # Define events to log and the arguments to log for each event.
+    # For more information about these events and what they mean, see https://peps.python.org/pep-0578/#suggested-audit-hook-locations
+    # For the full list of auditable events, see https://docs.python.org/3/library/audit_events.html
+    # Define this variable locally so it can't be modified by other modules.
+    #
+    # Events from the suggested audit hook locations in PEP 578 that we aren't logging:
+    #
+    # compile and import
+    #     Detects when code is being compiled or imported.
+    #
+    #     Why we aren't logging:
+    #     Logging as part of regular imports is too noisy.
+    #
+    # sys._getframe
+    #     Detects when code is accessing frames directly.
+    #
+    #     Why we aren't logging:
+    #     The logging module itself calls getFrame
+    #     (see https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L170)
+    #     so logging on this event would trigger an infinite loop unless
+    #     logging._srcfile is set to None
+    #     See https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L179-L189
+    #
+    # object.__getattr__
+    #     Detect access to restricted attributes. This event is raised for any built-in members
+    #     that are marked as restricted, and members that may allow bypassing imports.
+    #
+    #     Why we aren't logging:
+    #     The logging module itself accesses frame.f_code.co_filename
+    #     (see https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L202)
+    #     which calls object.__getattr__ under the hood
+    #     (see https://github.com/python/cpython/blob/main/Objects/frameobject.c#L89-L95)
+    #
+
+    EVENTS_TO_LOG = {
+        # Detect when new audit hooks are being added.
+        "sys.addaudithook": (),
+        # Detects any attempt to set the open_code hook.
+        "cpython.PyFile_SetOpenCodeHook": (),
+        # Detect dynamic execution of code objects. This only occurs for explicit
+        # calls, and is not raised for normal function invocation.
+        "exec": ("code_object",),
+        # Detect when a file is about to be opened. path and mode are the usual
+        # parameters to open if available, while flags is provided instead of
+        # mode in some cases.
+        "io.open": ("path", "mode", "flags"),
```

Comment:
Going by observed behaviour, some of these have different names. The list in https://peps.python.org/pep-0578/#suggested-audit-hook-locations is not right. This one is correct https://docs.python.org/3/library/audit_events.html

Examples:
`io.open` → `open`
`socket.address` → `socket.connect`, `socket.getaddrinfo`, `socket.gethostname`, `socket.bind`

Here's the full list we've observed in the last month on pfml:

```
array.__new__
fcntl.fcntl
marshal.dumps
object.__delattr__
open
os.fork
os.kill
os.mkdir
os.putenv
os.remove
os.rename
os.rmdir
shutil.rmtree
socket.bind
socket.connect
socket.getaddrinfo
socket.gethostname
subprocess.Popen
sys.excepthook
tempfile.mkdtemp
tempfile.mkstemp
```

Reply:
interesting. that leads me to realize that the existing unit tests aren't testing the right thing, since the unit test doesn't catch this. we should probably update the unit test to actually call some of underlying functions rather than directly call `handle_audit_event`